src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -0,0 +1,349 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal interface ITypeImportCompletionService : IWorkspaceService
+    {
+        void GetAccessibleTopLevelTypesFromPEReference(
+            Solution solution,
+            Compilation compilation,
+            PortableExecutableReference peReference,
+            Action<CompletionItem> handleAccessibleItem,
+            CancellationToken cancellationToken);
+
+        Task GetAccessibleTopLevelTypesFromCompilationReferenceAsync(
+            Solution solution,
+            Compilation compilation,
+            CompilationReference compilationReference,
+            Action<CompletionItem> handleAccessibleItem,
+            CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Get all the top level types from given project. This method is intended to be used for 
+        /// getting types from source only, so the project must support compilation. 
+        /// For getting types from PE, use <see cref="GetAccessibleTopLevelTypesFromPEReference"/>.
+        /// </summary>
+        Task GetAccessibleTopLevelTypesFromProjectAsync(
+            Project project,
+            Action<CompletionItem> handleAccessibleItem,
+            CancellationToken cancellationToken);
+    }
+
+    [ExportWorkspaceServiceFactory(typeof(ITypeImportCompletionService), ServiceLayer.Editor), Shared]
+    internal sealed class TypeImportCompletionService : IWorkspaceServiceFactory
+    {
+        private readonly ConcurrentDictionary<string, ReferenceCacheEntry> _peItemsCache
+            = new ConcurrentDictionary<string, ReferenceCacheEntry>();
+
+        private readonly ConcurrentDictionary<ProjectId, ReferenceCacheEntry> _projectItemsCache
+            = new ConcurrentDictionary<ProjectId, ReferenceCacheEntry>();
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            var cacheService = workspaceServices.GetService<IWorkspaceCacheService>();
+            if (cacheService != null)
+            {
+                cacheService.CacheFlushRequested += OnCacheFlushRequested;
+            }
```

Comment:
is this only called once? just making sure you're ont going to end up with thousands of subscriptions here every time you create a completoin list.

Reply:
I'd hope a workspace service is not created repeatedly every time someone tries to get one :)
Are you suggesting this might not be the case?

Reply:
can you debug through just to be sure :)  i'm just being paranoid.

Reply:
Confirmed.

Reply:
It's probably fine to have the service available, as long as only VS workspace ended up using it? Is it possible to limit the availability to certain types of workspace?